// cypress-qase-reporter/dist/metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const manager_1 = require("./metadata/manager");
module.exports = function (on) {
    on('task', {
        qaseTitle(value) {
            manager_1.MetadataManager.setTitle(value);
            return null;
        },
    });
    on('task', {
        qaseFields(value) {
            manager_1.MetadataManager.setFields(value);
            return null;
        },
    });
    on('task', {
        qaseIgnore() {
            manager_1.MetadataManager.setIgnore();
            return null;
        },
    });
    on('task', {
        qaseParameters(value) {
            manager_1.MetadataManager.setParameters(value);
            return null;
        },
    });
    on('task', {
        qaseGroupParameters(value) {
            manager_1.MetadataManager.setGroupParams(value);
            return null;
        },
    });
    on('task', {
        qaseSuite(value) {
            manager_1.MetadataManager.setSuite(value);
            return null;
        },
    });
    on('task', {
        qaseComment(value) {
            manager_1.MetadataManager.setComment(value);
            return null;
        },
    });
    on('task', {
        qaseStepStart(value) {
            manager_1.MetadataManager.addStepStart(value);
            return null;
        },
    });
    on('task', {
        qaseStepEnd(value) {
            manager_1.MetadataManager.addStepEnd(value);
            return null;
        },
    });
    on('task', {
        qaseAttach(value) {
            manager_1.MetadataManager.addAttach(value);
            return null;
        },
    });
};
