// qase-javascript-commons/dist/config/config-validation-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configValidationSchema = void 0;
const options_1 = require("../options");
const writer_1 = require("../writer");
/**
 * @type {JSONSchemaType<ConfigType>}
 */
exports.configValidationSchema = {
    type: 'object',
    properties: {
        mode: {
            type: 'string',
            enum: [options_1.ModeEnum.report, options_1.ModeEnum.testops, options_1.ModeEnum.off],
            nullable: true,
        },
        fallback: {
            type: 'string',
            enum: [options_1.ModeEnum.report, options_1.ModeEnum.testops, options_1.ModeEnum.off],
            nullable: true,
        },
        debug: {
            type: 'boolean',
            nullable: true,
        },
        environment: {
            type: 'string',
            nullable: true,
        },
        captureLogs: {
            type: 'boolean',
            nullable: true,
        },
        rootSuite: {
            type: 'string',
            nullable: true,
        },
        testops: {
            type: 'object',
            nullable: true,
            properties: {
                api: {
                    type: 'object',
                    nullable: true,
                    properties: {
                        token: {
                            type: 'string',
                            nullable: true,
                        },
                        host: {
                            type: 'string',
                            nullable: true,
                        },
                        headers: {
                            type: 'object',
                            nullable: true,
                            additionalProperties: false,
                            patternProperties: {
                                '^.*$': {
                                    type: 'string',
                                },
                            },
                        },
                        retries: {
                            type: 'number',
                            nullable: true,
                        },
                        retryDelay: {
                            type: 'number',
                            nullable: true,
                        },
                    },
                },
                project: {
                    type: 'string',
                    nullable: true,
                },
                uploadAttachments: {
                    type: 'boolean',
                    nullable: true,
                },
                run: {
                    type: 'object',
                    nullable: true,
                    properties: {
                        id: {
                            type: 'number',
                            nullable: true,
                        },
                        title: {
                            type: 'string',
                            nullable: true,
                        },
                        description: {
                            type: 'string',
                            nullable: true,
                        },
                        complete: {
                            type: 'boolean',
                            nullable: true,
                        },
                    },
                },
                plan: {
                    type: 'object',
                    nullable: true,
                    properties: {
                        id: {
                            type: 'number',
                            nullable: true,
                        },
                    },
                },
                batch: {
                    type: 'object',
                    nullable: true,
                    properties: {
                        size: {
                            type: 'number',
                            nullable: true,
                        },
                    },
                },
                defect: {
                    type: 'boolean',
                    nullable: true,
                },
                useV2: {
                    type: 'boolean',
                    nullable: true,
                },
            },
        },
        report: {
            type: 'object',
            nullable: true,
            properties: {
                driver: {
                    type: 'string',
                    enum: [writer_1.DriverEnum.local],
                    nullable: true,
                },
                connections: {
                    type: 'object',
                    nullable: true,
                    properties: {
                        [writer_1.DriverEnum.local]: {
                            type: 'object',
                            nullable: true,
                            properties: {
                                path: {
                                    type: 'string',
                                    nullable: true,
                                },
                                format: {
                                    type: 'string',
                                    enum: [writer_1.FormatEnum.json, writer_1.FormatEnum.jsonp],
                                    nullable: true,
                                },
                            },
                        },
                    },
                },
            },
        },
    },
};
